diff --git hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
index 08d3da5..621025e 100644
--- hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
+++ hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java
@@ -170,7 +170,7 @@ public void run(Path path) throws IOException {
       "\t\tThe directory quota is a long integer that puts a hard limit\n" +
       "\t\ton the number of names in the directory tree\n" +
       "\t\tFor each directory, attempt to set the quota. An error will be reported if\n" +
-      "\t\t1. N is not a positive integer, or\n" +
+      "\t\t1. quota is not a positive integer, or\n" +
       "\t\t2. User is not an administrator, or\n" +
       "\t\t3. The directory does not exist or is a file.\n" +
       "\t\tNote: A quota of 1 would force the directory to remain empty.\n";
@@ -271,7 +271,7 @@ public void run(Path path) throws IOException {
       "\t\tQuota can also be specified with a binary prefix for terabytes,\n" +
       "\t\tpetabytes etc (e.g. 50t is 50TB, 5m is 5MB, 3p is 3PB).\n" + 
       "\t\tFor each directory, attempt to set the quota. An error will be reported if\n" +
-      "\t\t1. N is not a positive integer, or\n" +
+      "\t\t1. quota is not a positive integer, or\n" +
       "\t\t2. user is not an administrator, or\n" +
       "\t\t3. the directory does not exist or is a file.\n" +
       "\t\tThe storage type specific quota is set when -storageType option is specified.\n";
diff --git hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testHDFSConf.xml hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testHDFSConf.xml
index 086bbe9..627f0e9 100644
--- hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testHDFSConf.xml
+++ hadoop-hdfs-project/hadoop-hdfs/src/test/resources/testHDFSConf.xml
@@ -15868,7 +15868,7 @@
         </comparator>
         <comparator>
           <type>RegexpComparator</type>
-          <expected-output>^( |\t)*1. N is not a positive integer, or( )*</expected-output>
+          <expected-output>^( |\t)*1. quota is not a positive integer, or( )*</expected-output>
         </comparator>
         <comparator>
           <type>RegexpComparator</type>
